custom_components/elasticsearch/es_gateway_8.py

```diff
@@ -0,0 +1,377 @@
+        except:  # noqa: E722
+            self._previous_ping = False
+
+            self._logger.debug("Error pinging Elasticsearch", exc_info=True)
+
+            return False
+        else:
+            self._previous_ping = True
+
+            return True
+
+    @async_log_enter_exit_debug
+    async def has_privileges(self, privileges) -> dict:
+        """Check if the user has the required privileges."""
+        with self._error_converter(msg="Error checking user privileges"):
+            response = await self.client.security.has_privileges(**privileges)
+
+        return self._convert_response(response)
+
+    @async_log_enter_exit_debug
+    async def get_index_template(self, name, ignore: list[int] | None = None) -> dict:
+        """Retrieve an index template."""
+        with self._error_converter(msg="Error retrieving index template"):
+            options = {}
+            if ignore:
+                options["ignore_status"] = ignore
+            response = await self.client.options(**options).indices.get_index_template(name=name)
+
+        return self._convert_response(response)
+
+    @async_log_enter_exit_debug
+    async def put_index_template(self, name, body) -> dict:
+        """Create an index template."""
+        with self._error_converter(msg="Error creating index template"):
+            response = await self.client.indices.put_index_template(name=name, **body)
+
+        return self._convert_response(response)
+
+    @async_log_enter_exit_debug
+    async def get_datastreams(self, datastream: str) -> dict:
+        """Retrieve datastreams."""
+        with self._error_converter(msg="Error retrieving datastreams"):
+            response = await self.client.indices.get_data_stream(name=datastream)
+
+        return self._convert_response(response)
+
+    @async_log_enter_exit_debug
+    async def rollover_datastream(self, datastream: str) -> dict:
+        """Rollover an index."""
+        with self._error_converter(msg="Error rolling over datastream"):
+            response = await self.client.indices.rollover(alias=datastream)
+
+        return self._convert_response(response)
+
+    @async_log_enter_exit_debug
+    async def bulk(self, actions: AsyncGenerator[dict[str, Any], Any]) -> None:
+        """Perform a bulk operation."""
+
+        with self._error_converter("Error performing bulk operation"):
+            count = 0
+            okcount = 0
+            errcount = 0
+            async for ok, result in async_streaming_bulk(
+                client=self.client,
+                actions=actions,
+                raise_on_error=False,
+                yield_ok=True,
+            ):
+                count += 1
+                action, outcome = result.popitem()
+                if not ok:
+                    errcount += 1
+                    self._logger.error("failed to %s document %s", action, outcome)
+                else:
+                    okcount += 1
+
+            if count > 0 or errcount == 0:
```

Comment:
I'm not sure this check is working correctly. I see a bunch of entries like this in my logs:
> 2024-08-05 14:00:01.305 INFO (MainThread) [custom_components.elasticsearch-httpsc88b3debf0fe4ce992f2473d8d43f5f1esuswest2awselasticcloud] Successfully published 0 documents

Should this be an `and` instead of an `or`?


```suggestion
            if count > 0 and errcount == 0:
```

Reply:
Thanks!